beacon-chain/sync/data_column_sidecars.go

```diff
@@ -0,0 +1,854 @@
+		}
+		requestedColumns = append(requestedColumns, reconstructedColumns[requestedIndex])
+	}
+
+	return requestedColumns, nil
+}
+
+// categorizeIndices separates indices into those that need to be queried from peers
+// and those that are already stored.
+func categorizeIndices(storage filesystem.DataColumnStorageReader, blockRoot [fieldparams.RootLength]byte, indices []uint64) (map[uint64]bool, map[uint64]bool) {
+	indicesToQuery := make(map[uint64]bool, len(indices))
+	indicesStored := make(map[uint64]bool, len(indices))
+
+	allStoredIndices := storage.Summary(blockRoot).Stored()
+	for _, index := range indices {
+		if allStoredIndices[index] {
+			indicesStored[index] = true
+			continue
+		}
+		indicesToQuery[index] = true
+	}
+
+	return indicesToQuery, indicesStored
+}
+
+// tryRequestingColumnsFromPeers attempts to request missing data column sidecars from connected peers.
+// It explores the connected peers to find those that are expected to custody the requested columns
+// and returns only when all requested columns are either retrieved or have been tried to be retrieved
+// by all possible peers.
+// Returns a map of block roots to their verified read-only data column sidecars and a map of block roots to indices of
+// still missing sidecars after all attempts.
+func tryRequestingColumnsFromPeers(
+	p DataColumnSidecarsParams,
+	roBlocks []blocks.ROBlock,
+	slotsWithCommitments map[primitives.Slot]bool,
+	indicesByRootToQuery map[[fieldparams.RootLength]byte]map[uint64]bool,
+) (map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, error) {
+	// Create a new random source for peer selection.
+	randomSource := rand.NewGenerator()
+
+	// Copy the requested indices to avoid modifying the original map.
+	missingIndicesByRoot := make(map[[fieldparams.RootLength]byte]map[uint64]bool, len(indicesByRootToQuery))
+	for root, indices := range indicesByRootToQuery {
+		missingIndicesByRoot[root] = make(map[uint64]bool, len(indices))
+		for index := range indices {
+			missingIndicesByRoot[root][index] = true
+		}
+	}
+
+	// Compute slots by block root.
+	slotByRoot := computeSlotByBlockRoot(roBlocks)
+
+	// Determine all sidecars each peers are expected to custody.
+	connectedPeersSlice := p.P2P.Peers().Connected()
+	connectedPeers := make(map[goPeer.ID]bool, len(connectedPeersSlice))
+	for _, peer := range connectedPeersSlice {
+		connectedPeers[peer] = true
+	}
+
+	indicesByRootByPeer, err := computeIndicesByRootByPeer(p.P2P, slotByRoot, missingIndicesByRoot, connectedPeers)
+	if err != nil {
+		return nil, errors.Wrap(err, "explore peers")
+	}
+
+	verifiedColumnsByRoot := make(map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn)
+	for len(missingIndicesByRoot) > 0 && len(indicesByRootByPeer) > 0 {
+		count := 0
+		for _, indices := range missingIndicesByRoot {
+			count += len(indices)
```

Comment:
count is incremented here but i don't see it used for verification or doing something else